### PR TITLE
Prereqs for cartesian rewrite

### PR DESCRIPTION
**Pull request by wemrysi**
The remainder after quotienting out the cartesian rewrite.

Depends on #4060.

### CODE REVIEW

**Comment by djspiewak** [comment]
Published as `106.0.1-5d42bcc`

**Comment by wemrysi** [comment]
Closing in favor of #4063

File: qscript/src/test/scala/quasar/qscript/MapFuncCoreSpec.scala [hunk 1]
```diff
@@ -284,6 +284,48 @@ final class MapFuncCoreSpec extends Qspec with TTypes[Fix] with TreeMatchers {
       normalize(expr) must beTreeEqual(expect)
     }
 
+    "project key from consequent when projecting from a Cond with an undefined alternate" >> {
```
Comment (alissapajer) [inline qscript/src/test/scala/quasar/qscript/MapFuncCoreSpec.scala]: What motivated the `Cond`-related changes?

File: qsu/src/test/scala/quasar/qsu/MinimizeAutoJoinsSpec.scala [hunk 2]
```diff
@@ -50,8 +50,9 @@ import scalaz.{\/, \/-, EitherT, Free, Need, StateT}
 import scalaz.std.anyVal._
 import scalaz.syntax.applicative._
 import scalaz.syntax.either._
-import scalaz.syntax.tag._
+// import scalaz.syntax.show._
```
Comment (alissapajer) [inline qsu/src/test/scala/quasar/qsu/MinimizeAutoJoinsSpec.scala]: Commented out import.

**Comment by alissapajer** [comment]
Oh, that's why the diff still looked... big. I was commented on the old PR :) Moving over to the new one...

File: qsu/src/test/scala/quasar/qsu/MinimizeAutoJoinsSpec.scala (hunk 2 above)
Comment (djspiewak) [reply qsu/src/test/scala/quasar/qsu/MinimizeAutoJoinsSpec.scala]: I added that. It's intentional, just like how I often add `// import scala._, Predef._` at the top of files. The reason being that it's insanely annoying to constantly have to re-add and then re-remove this line any time I need to debug an issue.

**Pull request closed without merging**